lib-api/src/main/java/org/ergoplatform/appkit/babelfee/BabelFeeOperations.java

```diff
@@ -0,0 +1,259 @@
+            return txB;
+        });
+    }
+
+    public static UnsignedTransaction cancelBabelFeeContract(BoxOperations boxOperations, InputBox babelBox) {
+        BabelFeeBoxState babelFeeBoxState = new BabelFeeBoxState(babelBox);
+
+        Address creatorAddress = Address.fromSigmaBoolean(babelFeeBoxState.getBoxCreator().getSigmaBoolean(),
+            boxOperations.getBlockchainContext().getNetworkType());
+
+        // if we can't pay tx fee from the babel box content, we need to load another box
+        List<InputBox> boxesToSpend;
+        if (babelFeeBoxState.getValue() < boxOperations.getFeeAmount() + Parameters.MinChangeValue)
+            boxesToSpend = boxOperations.loadTop();
+        else
+            boxesToSpend = new ArrayList<>();
+
+        boxesToSpend.add(0, babelBox);
+
+        UnsignedTransactionBuilder txB = boxOperations.getBlockchainContext().newTxBuilder();
+
+        OutBoxBuilder outBoxBuilder = txB.outBoxBuilder()
+            .value(Math.max(Parameters.MinChangeValue, babelFeeBoxState.getValue() - boxOperations.getFeeAmount()))
+            .contract(creatorAddress.toErgoContract());
+
+        if (!babelBox.getTokens().isEmpty())
+            outBoxBuilder.tokens(babelBox.getTokens().get(0));
+
+        return txB.boxesToSpend(boxesToSpend)
+            .fee(boxOperations.getFeeAmount())
+            .sendChangeTo(creatorAddress)
+            .outputs(outBoxBuilder.build())
+            .build();
+
+    }
+
+    /**
+     * Tries to fetch a babel fee box from blockchain data source with the given unspent boxes loader
+     *
+     * @param ctx       current blockchain context
+     * @param loader    loader to receive unspent boxes
+     * @param tokenId   tokenId offered to swap
+     * @param feeAmount nanoerg amount needed to swap
+     * @return babel fee box satisfying the needs, or null if none available
+     */
+    @Nullable
+    public static InputBox findBabelFeeBox(BlockchainContext ctx, BoxOperations.IUnspentBoxesLoader loader, ErgoId tokenId, long feeAmount) {
+        ErgoContract contractForToken = new BabelFeeBoxContract().getContractForToken(tokenId, ctx.getNetworkType());
+        Address address = contractForToken.toAddress();
+        loader.prepare(ctx, Collections.singletonList(address), feeAmount, new ArrayList<>());
+
+        int page = 0;
+        List<InputBox> inputBoxes = loader.loadBoxesPage(ctx, address, 0);
+
+        InputBox returnBox = null;
+        long pricePerToken = Long.MAX_VALUE;
+
+        while (!inputBoxes.isEmpty() && returnBox == null) {
+
+            // find the cheapest box satisfying our fee amount needs
+            for (InputBox inputBox : inputBoxes) {
+                try {
+                    BabelFeeBoxState babelFeeBoxState = new BabelFeeBoxState(inputBox);
+                    if (babelFeeBoxState.getValueAvailableToBuy() > feeAmount && babelFeeBoxState.getPricePerToken() < pricePerToken)
+                        returnBox = inputBox;
```

Comment:
seems like it is returning the first suitable (not the cheapest)

Reply:
The loop is not left, so the code runs until its end and will find the cheapest in the list. But when a box is found, no further page is fetched from the loader to avoid unnecessary traffic. This method is just a suggestion and not connected to any other code so developers can come up with their own selection mode.

Reply:
Ok, it actually returns the last suitable on the page. It is cheapest only if inputBoxes are ordered. Is it the case? if yes, please comment about this `cheapest` logic.

Reply:
done

Reply:
Maybe actually implement the cheapest selection as a default strategy?

Reply:
added a new parameter that can be used to retrieve all boxes.